test/e2e/network/traffic_distribution.go

```diff
@@ -0,0 +1,319 @@
+	// have no zone-hints.
+	endpointSlicesHaveNoHints := gcustom.MakeMatcher(func(slices []discoveryv1.EndpointSlice) (bool, error) {
+		for _, slice := range slices {
+			for _, endpoint := range slice.Endpoints {
+				var ip string
+				if len(endpoint.Addresses) > 0 {
+					ip = endpoint.Addresses[0]
+				}
+				if endpoint.Hints != nil && len(endpoint.Hints.ForZones) != 0 {
+					return false, fmt.Errorf("endpoint with ip %v has hint %+v, want no hint", ip, endpoint.Hints)
+				}
+			}
+		}
+		return true, nil
+	})
+
+	// requestsFromClient returns a helper function to be used with
+	// gomega.Eventually(...). It fetches the logs from the clientPod and returns
+	// them in reverse-chronological order.
+	requestsFromClient := func(clientPod *corev1.Pod) any {
+		return func(ctx context.Context) (reverseChronologicalLogLines []string, err error) {
+			logs, err := e2epod.GetPodLogs(ctx, c, f.Namespace.Name, clientPod.Name, clientPod.Spec.Containers[0].Name)
+			if err != nil {
+				return nil, err
+			}
+			framework.Logf("Logs from client=%q:\n%v", clientPod.GetName(), logs)
+			logLines := strings.Split(logs, "\n")
+			slices.Reverse(logLines)
+			return logLines, nil
+		}
+	}
+
+	////////////////////////////////////////////////////////////////////////////
+	// Main test specifications.
+	////////////////////////////////////////////////////////////////////////////
+
+	ginkgo.When("Service has trafficDistribution=PreferClose", func() {
+		ginkgo.It("should route traffic to an endpoint that is close to the client", func(ctx context.Context) {
+
+			ginkgo.By("finding 3 zones with schedulable nodes")
+			allZonesSet, err := e2enode.GetSchedulableClusterZones(ctx, c)
+			framework.ExpectNoError(err)
+			if len(allZonesSet) < 3 {
+				framework.Failf("got %d zones with schedulable nodes, want atleast 3 zones with schedulable nodes", len(allZonesSet))
+			}
+			zones := allZonesSet.UnsortedList()[:3]
+
+			ginkgo.By(fmt.Sprintf("finding a node in each of the chosen 3 zones %v", zones))
+			nodeList, err := e2enode.GetReadySchedulableNodes(ctx, c)
+			framework.ExpectNoError(err)
+			nodeForZone := make(map[string]string)
+			for _, zone := range zones {
+				found := false
+				for _, node := range nodeList.Items {
+					if zone == node.Labels[corev1.LabelTopologyZone] {
+						found = true
+						nodeForZone[zone] = node.GetName()
+					}
+				}
+				if !found {
+					framework.Failf("could not find a node in zone %q; nodes=\n%v", zone, format.Object(nodeList, 1 /* indent one level */))
+				}
+			}
+
+			ginkgo.By(fmt.Sprintf("creating 1 pod each in 2 zones %v (out of the total 3 zones)", zones[:2]))
+			zoneForServingPod := make(map[string]string)
+			var servingPods []*corev1.Pod
+			servingPodLabels := map[string]string{"app": f.UniqueName}
+			for _, zone := range zones[:2] {
+				pod := e2epod.NewAgnhostPod(f.Namespace.Name, "serving-pod-in-"+zone, nil, nil, nil, "serve-hostname")
+				pod.Spec.NodeName = nodeForZone[zone]
```

Comment:
we don't want to shortcut the scheduler as it has undesired consequences and creates flakiness and unstability on the test suite, use

```
		nodeSelection := e2epod.NodeSelection{Name: clientNodeInfo.name}
		e2epod.SetNodeSelection(&pod.Spec, nodeSelection)
```

Reply:
Oh interesting.

Does this mean the documentation at https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes/#create-a-pod-that-gets-scheduled-to-specific-node is more of a hack? Or maybe it is outdated after we encountered issues with that approach. Just curious.

Sure I'll make the change.

Reply:
is a hack in the sense that you override the scheduler, you can do it, but then you deal with the consequences of doing manual scheduling

Reply:
Gotcha! Changed